### PR TITLE
Refactor checkout-core to Kotlin 🎉 

### PR DESCRIPTION
**Pull request by caiofaustino**

### CODE REVIEW

File: checkout-core/src/main/java/com/adyen/checkout/core/exception/ApiCallException.kt [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+/**
+ * Exception thrown when there is an issue with an internal API call.
+ */
+class ApiCallException : CheckoutException {
+    constructor(errorMessage: String) : super(errorMessage)
+    constructor(errorMessage: String, cause: Throwable?) : super(errorMessage, cause)
+
+    companion object {
+        private const val serialVersionUID = 4060450855496938503L
```
Comment (OscarSpruit) [inline checkout-core/src/main/java/com/adyen/checkout/core/exception/ApiCallException.kt:18]: Are the `serialVersionUID `s in these classes needed? I don't get how they are used

File: checkout-core/src/main/java/com/adyen/checkout/core/api/HttpUrlConnectionFactory.kt [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+package com.adyen.checkout.core.api
+
+import com.adyen.checkout.core.log.LogUtil
+import com.adyen.checkout.core.log.Logger
+import java.net.HttpURLConnection
+
+/**
+ * A factory that creates a URL connections using a secure socket encryption.
+ */
+internal class HttpUrlConnectionFactory private constructor() : BaseHttpUrlConnectionFactory() {
```
Comment (jreij) [inline checkout-core/src/main/java/com/adyen/checkout/core/api/HttpUrlConnectionFactory.kt:17]: This ca be an `object` maybe, and we can get rid of `INSTANCE`?

File: checkout-core/src/main/java/com/adyen/checkout/core/log/LogUtil.kt [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+ */
+package com.adyen.checkout.core.log
+
+import android.os.Build
+import com.adyen.checkout.core.exception.NoConstructorException
+
+/**
+ * Utility class with methods related to logs.
+ */
+class LogUtil private constructor() {
```
Comment (OscarSpruit) [inline checkout-core/src/main/java/com/adyen/checkout/core/log/LogUtil.kt:16]: Might as well make this an `object`

File: checkout-core/src/main/java/com/adyen/checkout/core/log/LogUtil.kt [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        fun getTag(): String = getTag(CHECKOUT_LOG_PREFIX)
+
+        /**
+         * Get the TAG to be used for logging by the calling class.
+         *
+         * @return A String to be used as TAG with the format "Prefix.ClassName"
+         */
+        // This could be used by merchants if they want to.
+        fun getTag(prefix: String): String {
+            val callerClass = simplifiedCallerClassName
```
Comment (OscarSpruit) [inline checkout-core/src/main/java/com/adyen/checkout/core/log/LogUtil.kt:37]: I was wondering if it's a good idea to use class names. With obfuscation the class names are not very useful, but maybe that is intended 🤷

File: checkout-core/src/main/java/com/adyen/checkout/core/log/LogUtil.kt [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        @JvmStatic
+        fun getTag(): String = getTag(CHECKOUT_LOG_PREFIX)
+
+        /**
+         * Get the TAG to be used for logging by the calling class.
+         *
+         * @return A String to be used as TAG with the format "Prefix.ClassName"
+         */
+        // This could be used by merchants if they want to.
+        fun getTag(prefix: String): String {
```
Comment (OscarSpruit) [inline checkout-core/src/main/java/com/adyen/checkout/core/log/LogUtil.kt:36]: This should also have `@JvmStatic`

File: checkout-core/src/main/java/com/adyen/checkout/core/api/Connection.kt [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+        var message: String? = null
+        if (errorBytes != null) {
+            message = errorBytes.toString(CHARSET)
+        }
+        return IOException(message)
+    }
+
+    @Throws(IOException::class)
+    private fun getUrlConnection(url: String, headers: Map<String, String>, httpMethod: HttpMethod): HttpURLConnection {
+        val urlConnection = HttpUrlConnectionFactory.INSTANCE.createHttpUrlConnection(url)
```
Comment (jreij) [inline checkout-core/src/main/java/com/adyen/checkout/core/api/Connection.kt:92]: Can be replaced with `return HttpUrlConnectionFactory.INSTANCE.createHttpUrlConnection(url).apply{ ... } `

File: checkout-core/src/main/java/com/adyen/checkout/core/log/Logger.kt [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+import androidx.annotation.IntDef
+import com.adyen.checkout.core.BuildConfig
+import com.adyen.checkout.core.exception.NoConstructorException
+
+/**
+ * Log manager for Checkout.
+ * Serves as a proxy managing what and how to log information.
+ */
+// Keeping method names to match the ones from Logcat
+class Logger private constructor() {
```
Comment (OscarSpruit) [inline checkout-core/src/main/java/com/adyen/checkout/core/log/Logger.kt:20]: Could be an `object`

File: checkout-core/src/main/java/com/adyen/checkout/core/model/ModelObject.kt [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+abstract class ModelObject : Parcelable {
+    override fun describeContents(): Int {
+        return Parcelable.CONTENTS_FILE_DESCRIPTOR
+    }
+
+    /**
+     * Interface that must be implemented and provided as a public SERIALIZER field that serializes the to and from a JSONObject.
+     * @param <T> The class that extends ModelObject to be serialized.
+     */
+    interface Serializer<T : ModelObject?> {
```
Comment (OscarSpruit) [inline checkout-core/src/main/java/com/adyen/checkout/core/model/ModelObject.kt:36]: Is the nullability intended?

File: checkout-core/src/main/java/com/adyen/checkout/core/model/ModelObject.kt [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+                    throw CheckoutException("Failed to create ModelObject from parcel. JSONObject is null.")
+                }
+            } catch (e: JSONException) {
+                throw CheckoutException("Failed to create ModelObject from parcel.", e)
+            }
+            return ModelUtils.deserializeModel(jsonObject, mClass)
+        }
+
+        override fun newArray(size: Int): Array<T> {
+            return java.lang.reflect.Array.newInstance(mClass, size) as Array<T>
```
Comment (OscarSpruit) [inline checkout-core/src/main/java/com/adyen/checkout/core/model/ModelObject.kt:71]: There must be a prettier way to do this in Kotlin 🤔

File: checkout-core/src/main/java/com/adyen/checkout/core/log/Logger.kt (hunk 7 above)
Comment (jreij) [inline checkout-core/src/main/java/com/adyen/checkout/core/log/Logger.kt:20]: Could be object instead

File: checkout-core/src/main/java/com/adyen/checkout/core/model/ModelUtils.kt [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+package com.adyen.checkout.core.model
+
+import com.adyen.checkout.core.exception.BadModelException
+import com.adyen.checkout.core.exception.NoConstructorException
+import org.json.JSONArray
+import org.json.JSONObject
+import java.lang.reflect.Modifier
+import java.util.*
+
+class ModelUtils private constructor() {
```
Comment (OscarSpruit) [inline checkout-core/src/main/java/com/adyen/checkout/core/model/ModelUtils.kt:17]: Can be an `object`

File: checkout-core/src/main/java/com/adyen/checkout/core/log/LogUtil.kt (hunk 3 above)
Comment (jreij) [inline checkout-core/src/main/java/com/adyen/checkout/core/log/LogUtil.kt:16]: Could be object instead

File: checkout-core/src/main/java/com/adyen/checkout/core/model/ModelUtils.kt [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+        /**
+         * Parse a [JSONObject] to a class that extends [ModelObject] using its [ModelObject.Serializer].
+         *
+         * @param jsonObject The object to be parsed.
+         * @param modelClass The class type to be parsed to.
+         * @param <T> The type o the ModelObject class to be parse to.
+         * @return The parsed object.
+         */
+        @JvmStatic
+        fun <T : ModelObject?> deserializeModel(jsonObject: JSONObject, modelClass: Class<T>): T {
```
Comment (OscarSpruit) [inline checkout-core/src/main/java/com/adyen/checkout/core/model/ModelUtils.kt:30]: Is the nullability intended in all these methods?

File: checkout-core/src/main/java/com/adyen/checkout/core/model/ModelUtils.kt (hunk 10 above)
Comment (jreij) [inline checkout-core/src/main/java/com/adyen/checkout/core/model/ModelUtils.kt:17]: Could be object

File: checkout-core/src/main/java/com/adyen/checkout/core/util/ParcelUtils.kt [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+ * This file is open source and available under the MIT license. See the LICENSE file for more info.
+ *
+ * Created by caiof on 17/12/2020.
+ */
+package com.adyen.checkout.core.util
+
+import android.os.Parcel
+import com.adyen.checkout.core.exception.NoConstructorException
+
+class ParcelUtils private constructor() {
```
Comment (OscarSpruit) [inline checkout-core/src/main/java/com/adyen/checkout/core/util/ParcelUtils.kt:13]: Can be an `object`

File: checkout-core/src/main/java/com/adyen/checkout/core/util/StringUtil.kt [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+ *
+ * This file is open source and available under the MIT license. See the LICENSE file for more info.
+ *
+ * Created by caiof on 17/12/2020.
+ */
+package com.adyen.checkout.core.util
+
+import com.adyen.checkout.core.exception.NoConstructorException
+
+class StringUtil private constructor() {
```
Comment (OscarSpruit) [inline checkout-core/src/main/java/com/adyen/checkout/core/util/StringUtil.kt:12]: Can be an `object`

Comment (jreij) [inline checkout-core/src/main/java/com/adyen/checkout/core/util/StringUtil.kt:12]: Could be object

File: checkout-core/src/main/java/com/adyen/checkout/core/log/LogUtil.kt (hunk 4 above)
Comment (caiofaustino) [reply checkout-core/src/main/java/com/adyen/checkout/core/log/LogUtil.kt:37]: Yeah this is a whole conversation on how to improve our logging structure.

File: checkout-core/src/main/java/com/adyen/checkout/core/model/ModelObject.kt (hunk 8 above)
Comment (caiofaustino) [reply checkout-core/src/main/java/com/adyen/checkout/core/model/ModelObject.kt:36]: No it wasn't, I was already seeing issue in the next refactors, good catch.

File: checkout-core/src/main/java/com/adyen/checkout/core/model/ModelObject.kt (hunk 9 above)
Comment (caiofaustino) [reply checkout-core/src/main/java/com/adyen/checkout/core/model/ModelObject.kt:71]: I tried to look but I couldn't find anything. If you do let me know.
Alternatively I have a plan to try to make all final `ModelObject`s data classes and maybe use [@Parcelize](https://developer.android.com/kotlin/parcelize). Just an idea for now.

File: checkout-core/src/main/java/com/adyen/checkout/core/model/ModelUtils.kt (hunk 11 above)
Comment (caiofaustino) [reply checkout-core/src/main/java/com/adyen/checkout/core/model/ModelUtils.kt:30]: Nope 👍

**Review by jreij — APPROVED** [review]

File: checkout-core/src/main/java/com/adyen/checkout/core/model/ModelObject.kt (hunk 9 above)
Comment (OscarSpruit) [reply checkout-core/src/main/java/com/adyen/checkout/core/model/ModelObject.kt:68]: Hmm the best I can think of is this, but it's a bit hacky. Type erasure is difficult here

```
@Suppress("UNCHECKED_CAST")
override fun newArray(size: Int): Array<T?> {
        return arrayOfNulls<Any?>(size) as Array<T?>
}
```

**Review by OscarSpruit — APPROVED** [review]

**Pull request merged**